src/adaXT/decision_tree/decision_tree.pyx

```diff
@@ -165,29 +179,194 @@ class DecisionTree(BaseModel):
         return np.asarray(self.predictor.predict_proba(X))
 
     def get_leaf_matrix(self, scale: bool = False) -> np.ndarray:
+        cdef:
+            int i, j
+            int[::1] indices
+            int n_node
+            int n_rows
+
         if not self.root:
             raise ValueError("The tree has not been trained before trying to predict")
 
         leaf_nodes = self.leaf_nodes
-        n_obs = self.n_obs
-
-        matrix = np.zeros((n_obs, n_obs))
         if (not leaf_nodes):  # make sure that there are calculated observations
-            return matrix
-        for node in leaf_nodes:
-            if scale:
-                n_node = node.indices.shape[0]
-                matrix[np.ix_(node.indices, node.indices)] = 1/n_node
-            else:
-                matrix[np.ix_(node.indices, node.indices)] = 1
+            raise ValueError("The tree has no leaf nodes")
 
+        n_rows = self.n_rows
+        matrix = np.zeros((n_rows, n_rows))
+        for node in leaf_nodes:
+            indices = node.indices
+            for i in indices:
+                for j in indices:
+                    if scale:
+                        n_node = len(node.indices)
+                        matrix[i, j] += 1/n_node
+                    else:
+                        matrix[i, j] += 1
         return matrix
 
     def predict_leaf_matrix(self, X: np.ndarray, scale: bool = False):
         if not self.predictor:
             raise ValueError("The tree has not been trained before trying to predict")
         return self.predictor.predict_leaf_matrix(X, scale)
 
+    def __remove_leaf_nodes(self) -> None:
+        cdef:
+            int i, n_nodes
+            object parent
+        n_nodes = len(self.leaf_nodes)
+        for i in range(n_nodes):
+            parent = self.leaf_nodes[i].parent
+            if parent.left_child == self.leaf_nodes[i]:
+                parent.left_child = None
+            else:
+                parent.right_child = None
+            self.leaf_nodes[i] = None
+
+    def __fit_new_leaf_nodes(self, X: np.ndarray, Y: np.ndarray, sample_weight:
+                             np.ndarray, indices: np.ndarray) -> None:
+        cdef:
+            int idx, n_obs, depth, cur_split_idx
+            double cur_threshold
+            object cur_node
+            int[::1] all_idx
+            int[::1] leaf_indices
+
+        refit_objs = []
+        n_obs = indices.shape[0]
+        for idx in indices:
+            cur_node = self.root
+            depth = 0
```

Comment:
I'm probably wrong but shouldn't the depth be 1, since this value is set on the children?

Reply:
I think that is correct.